### PR TITLE
chore: remove the tag after the commit has been merged

### PR DESCRIPTION
**Pull request by tibi77**
This PR: 
Shall unpublish the npm tags after the PR has been merged, thus solving #901. For testing I was thinking we can change line 5 to trigger the action when merging in another branch, but opened to other ideas.

### CODE REVIEW

**Comment by Dhaiwat10** [comment]
@tibi77 are there any other changes you are planning to add, or is this PR ready for review as it is?

File: .github/workflows/pr-cleanup.yaml
Comment (arboleya) [inline .github/workflows/pr-cleanup.yaml:26]: I thought we could not access `github.event.pull_request.number` when a workflow is triggered by a push on the `master` branch. Did I get it wrong?

> cc @Torres-ssf

Comment (Torres-ssf) [reply .github/workflows/pr-cleanup.yaml:26]: Yes @arboleya, I have faced an error because of this while trying to get the proper workflow settings to deploy our new docs to GitHub pages. 

It seems the `github.event.pull_request` property is only available when:
```yaml
name: Pull Request Workflow

on:
  pull_request:
    branches:
      - branchName
```

**Comment by tibi77** [comment]
> @tibi77 are there any other changes you are planning to add, or is this PR ready for review as it is?

I m not 100% about testing it yet, so I m not really ready for review yet. Still looking for some help in there

**Comment by tibi77** [comment]
@arboleya hey! it s my first PR to fuel and I do not really know how review requests work, also I m not sure why my actions are not working
image

I get this

**Comment by arboleya** [comment]
@tibi77 We're all set here, and this should be reviewed soon.

We'll let you know if there's anything else we need to include here.

Thank you for your contribution so far.

**Comment by arboleya** [comment]
@tibi77 It seems that each iteration of a PR will effectively create a new version under the same tag.

Will the approach used here remove all versions for the PR tag in question?

For [example](https://www.npmjs.com/package/fuels?activeTab=versions):

versions

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
@tibi77 Thank you for your interest in contributing to this project. 

I have one question...

File: .github/workflows/pr-cleanup.yaml
Comment (tibi77) [reply .github/workflows/pr-cleanup.yaml:14]: ohhh right

**Comment by tibi77** [comment]
@arboleya, from what I gather from the documentation of the [unpublish command](https://docs.npmjs.com/policies/unpublish#what-to-do-if-your-package-does-not-meet-the-unpublish-criteria) unpublish without a specific version will just unpublish the entire package history.

**Comment by Torres-ssf** [comment]
@tibi77 Given the seriousness of this Pull Request and the concerns raised by Anderson, I will need to conduct some tests to ensure that everything runs smoothly. Please allow me a few days to complete these tests, and I will share the results with you.

I appreciate your time and dedication in solving this issue and thank you for your understanding.

Best regards

**Comment by tibi77** [comment]
Thanks a lot, @Torres-ssf please share the testing steps as well :D

**Comment by Torres-ssf** [comment]
> Thanks a lot, @Torres-ssf please share the testing steps as well :D

Will do mate

File: .github/workflows/pr-cleanup.yaml
Comment (LuizAsFight) [inline .github/workflows/pr-cleanup.yaml:10]: I think it should clean if it's pull request is not merged

**Review by LuizAsFight — CHANGES_REQUESTED** [review]
small change requested

also as tags will be deleted after closing/merging, I think we should remove `if:false` condition in yaml file... making tag always published for opened prs.

File: .github/workflows/pr-cleanup.yaml
Comment (arboleya) [reply .github/workflows/pr-cleanup.yaml:10]: Yes, I agree. The NPM tag should be removed regardless of whether the PR was merged.

File: .github/workflows/pr-cleanup.yaml
Comment (arboleya) [inline .github/workflows/pr-cleanup.yaml:1]:
```suggestion
name: Unpublish Pull Request NPM tag
```

File: .github/workflows/pr-cleanup.yaml
Comment (arboleya) [inline .github/workflows/pr-cleanup.yaml:9]:
```suggestion
  unpublish-pr:
```

File: .github/workflows/pr-cleanup.yaml
Comment (arboleya) [inline .github/workflows/pr-cleanup.yaml:26]: I guess this could be trimmed to:

```suggestion
        run: npm --force unpublish pr-${{ github.event.pull_request.number }}
```

**Review by arboleya — CHANGES_REQUESTED** [review]

File: .github/workflows/pr-cleanup.yaml
Comment (tibi77) [reply .github/workflows/pr-cleanup.yaml:10]: this means that we are just going to remove this if

**Comment by arboleya** [comment]
@tibi77 Thank you for the changes; everything is looking great now. 🚀 

However, I'll wait to approve it after @Torres-ssf finishes his [tests and validations](https://github.com/FuelLabs/fuels-ts/pull/903#issuecomment-1514812141).

**Comment by tibi77** [comment]
thanks a lot @arboleya

File: .github/workflows/pr-cleanup.yaml
Comment (Torres-ssf) [inline .github/workflows/pr-cleanup.yaml:27]:
```suggestion
      - name: Get package versions for PR tag
        id: pr_versions
        run: |
          VERSIONS=$(npm view fuels --json)
          PR_VERSIONS=$(echo "$VERSIONS" | tr ',' '\n' | tr -d '[]"' | grep -F "pr-${{ github.event.pull_request.number }}" | xargs)
          echo "::set-output name=versions::$PR_VERSIONS"
          
      - name: Unpublish NPM PR Tag Versions
        if: steps.pr_versions.outputs.versions != ''
        run: |
          for VERSION in ${{ steps.pr_versions.outputs.versions }}
          do
            echo "Unpublishing version $VERSION"
            npm --force unpublish fuels@${VERSION}
          done
        env:
          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

File: .github/workflows/pr-cleanup.yaml
Comment (Torres-ssf) [inline .github/workflows/pr-cleanup.yaml:16]: This seems to be required and without it, an error is thrown

```suggestion
      - name: Checkout repository
        uses: actions/checkout@v2
      
      - name: CI Setup
        uses: ./.github/actions/ci-setup

```

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
@tibi77, I've set up a test project on NPM to replicate the process we plan to use here. With each PR interaction, a new version is published under the associated PR tag.

Unfortunately, it appears that we can't delete all published versions tied to a PR tag using just one command. When running:

```yaml
npm --force unpublish pr-${{ github.event.pull_request.number }}
```

It only removes the latest version associated with the tag, leaving the other versions intact.

I wasn't able to find a single-command solution during my research. However, I did come across an alternative approach:

1. Retrieve all versions associated with the PR tag from NPM.
2. Iterate through the list of versions and unpublish each one individually.

This approach should help us achieve the desired outcome.

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
If I understand correctly, @LuizAsFight previously requested the removal of the following if statement:

https://github.com/tibi77/fuels-ts/blob/33164c6bbe156d2f4a27f61930d3fb189b40f0db/.github/workflows/pr-release.yaml#L10-L11

As this PR aims to resolve the issue of unpublishing all versions related to a PR tag, we can now remove the if statement. By removing it, the `pr-release.yaml` workflow will begin publishing versions associated with the respective PRs.

**Comment by arboleya** [comment]
@Torres-ssf This can be done on another PR if that's the case. However, I'm not sure if it makes sense to enable this auto-publish for _all_ PRs since only a few of them will effectively need it.

File: .github/workflows/pr-cleanup.yaml
Comment (arboleya) [reply .github/workflows/pr-cleanup.yaml:27]: A quick note about the `::set-output` command:

> **Warning**: The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

The warning above comes from [here](https://github.com/FuelLabs/fuels-ts/actions/runs/4779747120/jobs/8497023538?pr=827#step:6:21).

> **Note** The [`pr-release.yaml`](https://github.com/FuelLabs/fuels-ts/blob/master/.github/workflows/pr-release.yaml) workflow also uses this command.

Comment (arboleya) [reply .github/workflows/pr-cleanup.yaml:27]: Follow-up issue for existent workflows:
 - https://github.com/FuelLabs/fuels-ts/issues/928

**Comment by LuizAsFight** [comment]
> However, I'm not sure if it makes sense to enable this auto-publish for _all_ PRs since only a few of them will effectively need it.

@arboleya I think it's worth it, otherwise
- we would need to be always changing code when need, also remember to remove before merging PR
- if it's not my PR, would need to unnecessarily insert commits in someone else PR
- for the user, it's handful to have this ready whenever we need, it's like `next` tag, we don't need every time, but when we need it's just there ready to use.

**Comment by tibi77** [comment]
I am unsure if I understand if we should remove the `if false` from the pr release in the PR as well @arboleya @Torres-ssf @LuizAsFight

**Comment by arboleya** [comment]
@tibi77 Let's focus this PR on the `pr-cleanup.yaml` workflow in isolation.

The other question is more delicate and must be dealt with later on.

File: .github/workflows/pr-cleanup.yaml
Comment (arboleya) [inline .github/workflows/pr-cleanup.yaml:37]: @Torres-ssf Thank you for all tests and validations that resulted in the suggested version's iteration.

However, this will not be enough because this is limited to the `fuels` package only, whereas _all_ workspaces' packages are getting published. I believe the steps for a complete solution should look something like this:

 1. Iterate over **_all_** workspaces present in `pnpm-workspaces.yaml`
 1. List **_all_** packages in each workspace
 1. Check which ones are being **_published_**
 1. Fetch **_all_** published versions for each one
 1. Finally, **_unpublish_** them all **_individually_**

The complications here tell me that this will be a very risky and challenging task.

In fact, it makes me question how feasible this is and if it's worth the trouble at the moment.

Let's think about this more carefully before we move on.

**Review by arboleya — CHANGES_REQUESTED** [review]

File: .github/workflows/pr-cleanup.yaml
Comment (tibi77) [reply .github/workflows/pr-cleanup.yaml:37]: I think we might be able to do something like this :
```
pnpm list -r --depth=-1 | awk '{ if($0 !~ /^[[:space:]]*$/) print $1}'
```
to get the list of names and then we need to iterate through them. but from what I've seen, some of the packages return an error for the npm view, so we need to be attentive of that as well.
(example forc-bin)
@Torres-ssf could you show some docs of how to set up the npm package myself, I'd really want to have a good testing platform before going further with this.

Comment (arboleya) [reply .github/workflows/pr-cleanup.yaml:37]: @tibi77 Great insight! 🚀 

Not _all_ packages are currently published, these should be marked as private:

https://github.com/FuelLabs/fuels-ts/blob/dadc5760549564480b7b580cb5c268030bf7dcd6/packages/forc-bin/package.json#L2

Comment (tibi77) [reply .github/workflows/pr-cleanup.yaml:37]: Ahhh makes sense now, ok, will try to hack something together :D

**Comment by tibi77** [comment]
hey! wrote the script. @Torres-ssf could you please help me regarding the testing of it :)
also I d be really happy if i can test this locally, the multiple versions thing... it is quite complex to set up as far as I gather :(

File: .github/workflows/pr-cleanup.yaml
Comment (tibi77) [reply .github/workflows/pr-cleanup.yaml]: I think I got it :)

**Comment by Torres-ssf** [comment]
> hey! wrote the script. @Torres-ssf could you please help me regarding the testing of it :) also I d be really happy if i can test this locally, the multiple versions thing... it is quite complex to set up as far as I gather :(

Hey @tibi77 I am deeply sorry for the late response.

To validate what we are trying to accomplish here, I've created this [dummy project here](https://github.com/Torres-ssf/dummy-test-npm-project-torres).

Unfortunately, I do not have a good doc or cheat sheet to point you right now. To properly execute these tests I have gone to a good amount of Google searches and have scrapped pieces of code/information from many different fonts while trying to modify them to our specific needs.

The most complicated/tricky parts when setting up this test project were:

1. Create the workflow that will generate the PR tag name and will publish it to NPM on each interaction on the PR (Needs to mimic the same PR tag name approach as `fuels-ts`)
2. Create the unpublish workflow and execute the tests.

You can validate my implementation for each case here:
- [pr-release.yaml](https://github.com/Torres-ssf/dummy-test-npm-project-torres/blob/master/.github/workflows/pr-release.yaml): This one is going to release the versions to NPM with a PR tag. The logic behind the PR tag name mimics the same output we would have on the PRs Tag that is being created on `fuels`. Similar to what @arboleya has shown [here](https://github.com/FuelLabs/fuels-ts/pull/903#issuecomment-1514661141).
- [pr-unpublish.yaml](https://github.com/Torres-ssf/dummy-test-npm-project-torres/blob/master/.github/workflows/pr-unpublish.yaml): This one test the work we are doing here.

Please let me know if you have any other questions related to this and my implementation of this dummy project.

I promise that this time I won't be taking too much time to answer you :sweat_smile: 

Good luck on your tests

**Comment by tibi77** [comment]
@Torres-ssf thanks a lot.
Just making sure, this is the correct [npm package](https://www.npmjs.com/package/dummy-test-npm-project-torres?activeTab=versions), right.

**Comment by Torres-ssf** [comment]
> @Torres-ssf thanks a lot. Just making sure, this is the correct [npm package](https://www.npmjs.com/package/dummy-test-npm-project-torres?activeTab=versions), right.

@tibi77  Exactly

**Comment by LuizAsFight** [comment]
FYI @pedronauck has implemented to unpublish from npm [here](https://github.com/FuelLabs/fuel-ui/pull/244), feel free to reuse useful parts of his code

**Comment by tibi77** [comment]
@Torres-ssf I think from what @arboleya  said in [here](https://github.com/FuelLabs/fuels-ts/pull/903#discussion_r1179641728) we need to set up another package as well so that we see if the current code is deleting those as well.

**Comment by tibi77** [comment]
@LuizAsFight do you happen to know how @pedronauck did the tests ?

**Comment by Torres-ssf** [comment]
> @Torres-ssf I think from what @arboleya said in [here](https://github.com/FuelLabs/fuels-ts/pull/903#discussion_r1179641728) we need to set up another package as well so that we see if the current code is deleting those as well.

@tibi77 Yes it seems so. I couldn't validate that yet. I've worked on these tests before @arboleya raised this concern.

**Comment by tibi77** [comment]
hey @Torres-ssf trying to push to that repo that you created and I get permission denied, could you please add me as a collaborator ?

**Comment by Torres-ssf** [comment]
@tibi77 Done!

**Comment by tibi77** [comment]
@Torres-ssf mate you are like a race car. thanks a lot. the github action for deploying to npm is just when a pr is merged?

**Comment by Torres-ssf** [comment]
> @Torres-ssf mate you are like a race car. thanks a lot. the github action for deploying to npm is just when a pr is merged?

No. When opening a new PR the workflow already starts deploying to NPM under the PR tag.

**Comment by tibi77** [comment]
Patched the set output one, 
quick question why does the name of the release action is 
```Release to @pr-<number> tag on npm / Release PR to...``` can we not take the PR number inside the action title?

**Comment by tibi77** [comment]
hey @Torres-ssf @arboleya @LuizAsFight 
kind reminder regarding this PR

**Pull request closed without merging**
